src/telemetry/battery_status.rs

```diff
@@ -0,0 +1,140 @@
+/*
+ * This file is part of Edgehog.
+ *
+ * Copyright 2022 SECO Mind Srl
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use serde::Serialize;
+use std::collections::HashMap;
+
+use crate::error::DeviceManagerError;
+use crate::telemetry::upower::device::{BatteryState, DeviceProxy, PowerDeviceType};
+use crate::telemetry::upower::upower::UPowerProxy;
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct BatteryStatus {
+    level_percentage: f64,
+    level_absolute_error: f64,
+    /// "Battery status string, any of: Charging, Discharging, Idle, EitherIdleOrCharging, Failure, Removed, Unknown",
+    status: String,
+}
+
+pub async fn get_battery_status() -> Result<HashMap<String, BatteryStatus>, DeviceManagerError> {
+    let connection = zbus::Connection::system().await?;
+
+    let upower = match UPowerProxy::new(&connection).await {
+        Ok(upower) => upower,
+        Err(_) => return Ok(HashMap::new()),
+    };
+
+    let devices = match upower.enumerate_devices().await {
+        Ok(devices) => devices,
+        Err(_) => return Ok(HashMap::new()),
+    };
+
+    let mut result = HashMap::new();
+    for device_path in devices {
+        let device = DeviceProxy::builder(&connection)
+            .path(device_path)?
+            .build()
+            .await?;
+
+        if device.power_device_type().await? == PowerDeviceType::Battery {
+            let device_state = device.state().await?;
+            let status = get_status(device_state, device.is_present().await?);
+
+            let level_absolute_error = get_accuracy(device_state);
+
+            let battery = BatteryStatus {
+                level_percentage: device.percentage().await.unwrap(),
+                level_absolute_error,
+                status,
+            };
+
+            result.insert(device.serial().await?, battery);
+        }
+    }
+
+    Ok(result)
```

Comment:
I thought of another approach for this that could reduce repetition without introducing back highly indented code: As you could have a `get_powet_devices() -> Result<Something>` function that chains the calls starting from `zbus::Connection::system` with the `?` operator and then the outer function just handles the `Ok/Err` case in a single place and either iterates on `power_devices` or returns an empty HashMap